text/3606-temporary-lifetimes-in-tail-expressions.md

```diff
@@ -0,0 +1,167 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+In the next edition, drop temporaries in tail expressions *before* dropping locals, rather than after.
+
+![A diagram showing a function with one let statement "let x = g();" and a tail expression "temp().h()"
+and a visualisation of how long x and temp live before and after this change.
+Before: x is created first, then temp is created, then x is dropped, then temp is dropped.
+After: x is created first, then temp is created, then temp is dropped, then x is dropped.
+](https://hackmd.io/_uploads/HyVB0FtkA.svg)
+
+# Motivation
+
+Temporaries in the tail expression in a block live longer than the block itself,
+so that e.g. `{expr;}` and `{expr}` can behave very differently.
+
+For example, this fails to compile:
+
+```rust
+// This fails to compile!
+fn f() -> usize {
+    let c = RefCell::new("..");
+    c.borrow().len() // ERROR!!!
+}
+```
+
+The temporary `std::cell::Ref` created in the tail expression will be dropped
+after the local `RefCell` is dropped, resulting in a lifetime error.
+
+This leads to having to add seemingly unnecessary extra `let` statements
+or having to add seemingly unnecessary semicolons:
+
+```rust
+fn main() {
+    let c = std::cell::RefCell::new(123);
+
+    if let Ok(mut b) = c.try_borrow_mut() {
+        *b = 321;
+    }; // <-- Error if you remove the semicolon!
+}
+```
+
+Both of these examples will compile fine after the proposed change.
+
+# Guide-level explanation
+
+Temporaries are normally dropped at the end of the statement.
+
+The tail expression of a block
+(such as a function body, if/else body, match arm, block expression, etc.)
+is not a statement, so has its own rule:
+
+- Starting in Rust 2024,
+  temporaries in tail expressions are dropped after evaluating the tail expression,
+  but before dropping any local variables of the block.
+
+For example:
+
+```rust
+fn f() -> usize {
+    let c = RefCell::new("..");
+    c.borrow().len() // Ok in Rust 2024
+}
+```
+
+The `.borrow()` method returns a (temporary) `Ref` object that borrows `c`.
+Starting in Rust 2024, this will compile fine,
+because the temporary `Ref` is dropped before dropping local variable `c`.
+
+# Reference-level explanation
+
+For blocks/bodies/arms whose `{}` tokens come from Rust 2024 code,
+temporaries in the tail expression will be dropped *before* the locals of the block are dropped.
+
+# Breakage
+
+It is very hard to come up with examples that will stop compiling.
+Such code will involve a tail expression that injects a borrow to a temporary
+into an already existing local variable that borrows it on drop.
+
+For example:
+
+```rust
+fn why_would_you_do_this() -> bool {
+    let mut x = None;
+    // Make a temporary `RefCell` and put a `Ref` that borrows it in `x`.
+    x.replace(RefCell::new(123).borrow()).is_some()
+}
+```
+
+We expect such patterns to be very rare in real world code.
```

Comment:
````suggestion
We expect such patterns to be very rare in real world code. Another example of code that will not compile is the following. This example would error in Rust 2024 because the `Some(tmp.len())` temporary will be dropped when the block complete, whereas in Rust 2021 it is dropped at the end of the statement.

```rust
fn why_would_you_do_this() -> usize {
    *foo(
        {
            let tmp = vec![44];
            Some(tmp.len()).as_ref() // Option<&u32>
        },
        &(22+44),
    )
}

fn foo<'a, T>(
    x: Option<&'a T>,
    y: &'a T,
) -> &'a T {
    x.unwrap_or(y)
}
```
````

Reply:
@nikomatsakis I'm trying to simplify the example for clarity. Am I right that this simpler example below demonstrates the exact same thing as your example?

```rust
fn example() -> usize {
    { String::new().as_str() }.len() // compiles today, will error after this RFC.
}
```

(With the new rules, the temporary `String::new()` will be dropped before `.len()` is executed on the `&str` that borrows that temporary String.)

Reply:
I added this example and added an option to the edition migration section to address such simple cases (by e.g. replacing `{}` by `()`).

Reply:
Does this mean that if instead of `as_str` we use a raw pointer method, then this would be code that is fine today but has Undefined Behavior after the RFC?

Reply:
Yes. I have added an unresolved question for that. (See also https://github.com/rust-lang/rfcs/pull/3606#issuecomment-2047999540)